py-rse/errors.Rmd

```diff
@@ -155,87 +154,134 @@ except ArithmeticError as error:
-        print('No log directory {}'.format(LOG_DIR), file=sys.stderr)
+```text
+Traceback (most recent call last):
+  File "bin/collate.py", line 27, in <module>
+    main(args)
+  File "bin/collate.py", line 17, in main
+    with open(file_name, 'r') as reader:
+FileNotFoundError: [Errno 2] No such file or directory: 'results/none.csv'
 ```
 
-With the second,
-we try the operation and catch the exception if one is raised:
+But what happens if we are trying to read an file that exists,
+but was not created by `countwords.py`?
 
-```python
-LOG_DIR = "/tmp/mylogdir"
 
-def write_message(filename, message):
-    try:
-        path = os.path.join(LOG_DIR, filename)
-        with open(path, 'a') as writer:
-            writer.write(message)
-    except OSError as error:
-        print('Unable to write log message to {}: {}'.format(path, error))
-```{: title="logging/ask_permission_not_forgiveness.py"}
-
-The first approach doesn't actually need exception handlers,
-except it does:
-programmers can never anticipate all of the things that can go wrong in the real world,
-so there should always be an `except` somewhere.
-Since that's the case,
-most programmers prefer the second approach.
-
-When we are raising exceptions ourselves,
-we should make the information as specific as possible.
-For example,
-suppose a program contains this:
+```shell
+$ python bin/collate.py Makefile
+```
+
+```text
+Traceback (most recent call last):
+  File "bin/collate.py", line 27, in <module>
+    main(args)
+  File "bin/collate.py", line 18, in main
+    update_counts(reader, word_counts)
+  File "bin/collate.py", line 10, in update_counts
+    for word, count in csv.reader(reader):
+ValueError: not enough values to unpack (expected 2, got 1)
+```
+
+This error is not that easy to understand.
+It would be useful if our programmed check that the input files were CSV-files,
+and if not,
+raised an error with a useful explanation to what went wrong.
+We could achieve this by wrapping the open statement in a `try/except` clause:
 
 ```python
-if (val is not None) and (min_val < val < max_val) and (val % 2 == 0) and (val >= prev_val):
-    # OK branch
-else:
-    raise ValueError('Bad value: {}'.format(val))
+for file_name in args.infiles:
+    try:
+        with open(file_name, 'r') as reader:
+            update_counts(reader, word_counts)
+    except ValueError as e:
+        print(f'{file_name} is not a CSV-file.')
+        print(f'ValueError: {e}')
+```
+
+```shell
+$ python bin/collate.py Makefile
 ```
 
-Knowing what value caused the error is helpful,
-but the message in the exception doesn't tell us why it was bad.
-Was it not between `min_val` and `max_val`?
-Was it less than `prev_val`?
-Whoever has to debug the problem will want to know,
-but breaking the test up into cases makes the code harder to read:
+```text
+Makefile is not a CSV-file.
+ValueError: not enough values to unpack (expected 2, got 1)
+```
+
+This is definitely more informative than before!
```

Comment:
No exclamation marks please.

Reply:
But it is so much more informative! ... removed